### PR TITLE
let NOTICES be double gzip wrapped to reduce on-disk installed space

### LINKED ISSUE #71102: Deduplicate NOTICES
On 1.24.0-8.0.pre.344, the `LicenseResult` collected from LicenseCollector https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/asset.dart#L322 produce a lot of redundancies. 

The uncompressed size of the file is relevant for https://github.com/flutter/flutter/issues/71098. 

A breakdown of occurrences per library shows:

```
  86 harfbuzz
  75 boringssl
  47 skia
  34 turbo
  34 libjpeg
  29 angle
  25 zlib
  22 icu
  19 freetype2
  17 dart
  12 fuchsia_sdk
   9 libwebp
   8 glfw
   4 rapidjson
   4 files
   4 base
   3 libcxx
   3 khronos
   3 double
   3 cpp
   3 conversion
   3 abseil
   2 xxhash
   2 tcmalloc
   2 skcms
   2 root_certificates
   2 ffx_spd
   1 wuffs
   1 wasmer
   1 vulkanmemoryallocator
   1 vector_math
   1 test_api
   1 term_glyph
   1 smhasher
   1 pkg
   1 libsdl
   1 flutter
   1 engine
   1 colorama
   1 clock
   1 charcode
   1 characters
   1 boolean_selector
   1 async
   1 StackWalker
```

Some examples of "different" entries:

```
skia

Copyright 2018 Google Inc.
...
```

```
skia

Copyright 2018 Google LLC
...
```

```
skia

Copyright 2018 Google LLC.
```

```
skia

Copyright 2018 Google, LLC
```

Do we need all these variations?

### PR DESCRIPTION
**Pull request by xster**
Fixes https://github.com/flutter/flutter/issues/71102. It doesn't quite deduplicate the contents of NOTICES while in transport, but just leaves it compressed on disk until it's actually read since it's an uncommonly used file. Reduces uncompressed (.ipa perspective, not this file's perspective) size from ~800kB before to ~60kB after. 

It leaves web as is since the client doesn't have dart:io or any "native" abilities to unzip. It also has a .Z extension instead of a .gz extension because there are logic baked in in gradle that traverses all the files of an APK and unzips the .gz files. Checked with the gradle team. Our usage here isn't circumventing any meaningful optimizations. It was just originally meant to avoid people thinking this reduces network size (which this doesn't and doesn't intend to) and costing unzip CPU cost (which doesn't affect this since it's uncommonly used). 

Added some new tests but https://github.com/flutter/flutter/pull/71899/files#diff-7e359e27713e3f080b4385c7e6b690e7aa3ab982c98855a088ed00e07bfaa6e5R16 actually ends up making this get tested everywhere. 

----------------------

For future rollers, this will break g3. Subclasses will need to implement the new signature to match.

### CODE REVIEW

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 1]
```diff
@@ -73,15 +79,25 @@ abstract class AssetBundle {
...
+    return compute(
+      unzip ? _utf8ZipDecode : _utf8Decode,
+      data,
+      debugLabel: 'UTF8 decode for "$key"',
+    );
+  }
+
+  static String _utf8ZipDecode(ByteData data) {
+    List<int> bytes = data.buffer.asUint8List();
+    bytes = gzip.decode(bytes);
```
Comment (xster) [inline packages/flutter/lib/src/services/asset_bundle.dart:96]: @yjbanov web question while Jonah is out. I can't do this on web right?

Comment (yjbanov) [reply packages/flutter/lib/src/services/asset_bundle.dart:96]: `dart:io` is not available. However, the server and browser are expected to compress all data exchanged over the network, so we can just skip the decoding here if we're on the web.

**Comment by jmagman** [comment]
`7za` isn't available on LUCI
https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20module_test/1747/overview
```
  "success": false,
  "reason": "Invalid argument(s): Cannot find executable for 7za."
```

You can see the powershell fallbacks here:
https://github.com/flutter/flutter/blob/23c7ee9debd0a0efaa07ec7165f51e5767df0992/bin/internal/update_dart_sdk.ps1#L73-L89

**Comment by xster** [comment]
Thanks for the tip. I ended up just unpacking the APK in dart instead of shelling out.

**Review by jmagman — APPROVED** [review]
The code LGTM. Can you give this a TestFlight run to confirm there's no unexpected consequences to publishing an app with that embedded file type?

**Comment by xster** [comment]
g3 rollers: subclasses must implement https://github.com/flutter/flutter/pull/71899/files#diff-1339eded5eeb35ab9fbc61051c530c73e307d4640328760c1af0eee36c142be2R172.

**Comment by renyou** [comment]
> g3 rollers: subclasses must implement https://github.com/flutter/flutter/pull/71899/files#diff-1339eded5eeb35ab9fbc61051c530c73e307d4640328760c1af0eee36c142be2R172.

Hi, @xster , actually I didn't see this comment until now. Could you please communicate on the internal channels next time so we can track this sooner? Thanks!

**Comment by xster** [comment]
@renyou will do. Though this one in particular was reverted, so no action needed in this case.

**Pull request merged**